docs/proposals/edgemesh-design.md

```diff
@@ -76,6 +76,113 @@ Add two endpoints definition:
 |ServiceBus|Target|Request Forward|
 |Router|Source|Resutst Forward, Service Discovery, Load Balance|
 
+<img src="../images/proposals/router-design.png">
+
+* R1: EdgeController watches on k8s apiserver for Services/Endpoints.
+* R2: EC get the service definition and create router rule definition and save in DB(sqlite).
+* R3: Router gets the rule definition from DB and initializes the listeners/handlers for source and target.
+* R4: Users/Applications create a request.
+* R5: Router receives the request and forwards it to the target based on rule.
+
+In Alpha1.0 only servicebus and eventbus will be supported.
+
+Since Router fetches rules from DB, in later versions it can be started as a different process than edge-controller. Also it enables router to retrieve all the rules after process restarts.
+### Router Low-level Design
+
+### Providers
+Providers can be classified into two types
+1) Source
+2) Target
+
+### Source Interfaces
+
+Each provider that wants to register as a Source should implement following interfaces.
+
+```go
+type Source interface {
+	// Name Returns name of the provider
+	Name() string
+	// RegisterListener is used to register listener for a rule with this source type
+	RegisterListener(rule model.Rule, res model.Map, handler func(interface{})) error
+	// UnregisterListener is used to unregister a listener when rule is deleted
+	UnregisterListener(rule model.Rule, res map[string]interface{})
+	// Callback is function for sending response/ACK if required for a request 
+	Callback(map[string]interface{})
+}
+```
+
+```go
+type SourceFactory interface {
+	// Type returns type of the source
+	Type() string
+	// GetSource returns Source implementation
+	GetSource(ep *model.Endpoint) Source
+}
+```
+
+### Target Interfaces
+
+Each provider that wants to register as a Target should implement following interfaces.
+
+```go
+type Target interface {
+	// Name returns name of the provider
+	Name() string
+	// Forward is used to forward the request to target 
+	Forward(map[string]interface{}, interface{}, func(map[string]interface{})) error
+}
+```
+
+```go
+type TargetFactory interface {
+	// Type returns type of the target
+	Type() string
+	// GetTarget returns Target implementation
+	GetTarget(ep *model.Endpoint) Target
+}
+```
+
+### Rule Definition
+
+```go
+// Rule define message translate from Endpoint to Endpoint
+type Rule struct {
```

Comment:
can we have comments on top of the member variables of the rule to explain what is it about ?

Reply:
Done, Thanks